include/openthread/link_metrics.h

```diff
@@ -0,0 +1,164 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *   This file defines the OpenThread Link Metrics API.
+ */
+
+#ifndef LINK_METRICS_H_
+#define LINK_METRICS_H_
+
+#include "openthread-core-config.h"
+
+#include <openthread/ip6.h>
+#include <openthread/message.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup linkmetrics   Link Metrics
+ *
+ * @brief
+ *   This module includes functions that control Link Metrics protocol.
+ *
+ * @{
+ *
+ */
+
+enum
+{
+    OT_LINK_METRICS_TYPE_ID_MAX_COUNT = 4, ///< Max TypeIdFlags count in a Link Metrics query
+};
+
+/**
+ * This enumeration defines Link Metrics ID.
+ *
+ */
+typedef enum otLinkMetricsId
+{
+    OT_LINK_METRICS_PDU_COUNT = 0, ///< Layer 2 PDUs Received
+    OT_LINK_METRICS_LQI       = 1, ///< Layer 2 LQI
+    OT_LINK_METRICS_MARGIN    = 2, ///< Link Margin - RSSI margin above noise floor
+    OT_LINK_METRICS_RSSI      = 3, ///< RSSI
+} otLinkMetricsId;
+
+/**
+ * This enumeration defines Link Metrics type.
+ *
+ */
+typedef enum otLinkMetricsType
+{
+    OT_LINK_METRICS_METRIC_COUNT_SUMMATION            = 0, ///< Count/summation
+    OT_LINK_METRICS_METRIC_EXPONENTIAL_MOVING_AVERAGE = 1, ///< Exponential moving average
+} otLinkMetricsType;
+
+/**
+ * This structure represents Link Metrics Type ID Flags.
+ */
+typedef struct otLinkMetricsTypeIdFlags
```

Comment:
Agree to use a simpler API like this.

I found in test plan 7_2_1, the **L** bit is set by the user for `count`. So I think we need to let user to choose if **L** bit needs to be set for `count`. We may add a field `mPsduCountLBit` in `otLinkMetrics`. Thoughts? @librasungirl @abtink 

And I'm thinking the format of api in cli command: how should cli pass the information for `otLinkMetrics`. If we pass a hex byte in cli command and then copy it into a `otLinkMetrics`, I'm not sure if bit fields always work correctly within one byte. Shall we also use the 'Mask' style for `otLinkMetrics`? i.e. its member is just one byte, we fetch its fields by masks.

Reply:
What does it mean for requester to specify the `L` bit? How should it change the behavior?
Is there a test-plan where it is set and another where it is not set?

To me the `L` should be applicable when there is a value being reported (to indicate its length). So in the query/request message (which contains no metric value), I think the `L` bit should be just ignored. 

As @librasungirl  mentioned, having the request/query msg specify the `L` and the size of metric value, can lead to weird/invalid situations, e.g., if request say I want "psdu count"  as a `uint8_t`, how should the reporter handle it if the actual value is larger than 255 (not report it? cap it to 255?)

----

That said, if there is a need to add this somehow in the request I would suggest doing something like this:
```
typedef struct oLinkMetics
{
    bool mPsduCountUint8 : 1;   // PSDU count (with `uint8_t` value)
    bool mPsduCountUint32 : 1;  // PSDU count (with `uint16_t` value)
    bool mLqi : 1;
    bool mLinkMargin : 1;
    bool mRssi : 1;
} otLinkMetrics;
```
Or maybe `mPsduCountShort` and `mPsduCountLong` to indicate it. I would not recommend adding a `mPsduCountLBit` flag (since it would require user to be aware of what an `L` flag means - which is the protocol/spec detail).  

------

> And I'm thinking the format of api in cli command: how should cli pass the information for otLinkMetrics. If we pass a hex byte in cli command and then copy it into a otLinkMetrics, I'm not sure if bit fields always work correctly within one byte. Shall we also use the 'Mask' style for otLinkMetrics? i.e. its member is just one byte, we fetch its fields by masks.

For CLI I would suggest using character flags (like `mode` command which uses `rsdn`).

Reply:
`rsdn` are defined by Thread spec. If we want to use character flags, should we propose define it in the spec as well? Otherwise, it may be confusing. Thoughts?

Reply:
Currently, the THCI for using Link Metrics feature uses hex byte array to pass `TypeIdFlags`. I think it's indeed unnecessary and we can make the public API much simpler. (Using character flags instead of hex byte array).

The character flags here are only for public API usage(CLI command). But the spec doesn't have requirements for how should our public API looks like.  I think we can define our own character flags just for cli usage. (of course with good documentation)